### PR TITLE
Stargenerator Vertexing Options

### PR DESCRIPTION
**Pull request by klendathu2k**
Added several options for sampling "flat" vertex distributions:
1) Line along z-axis
2) Cylinder along z-axis
3) Elliptical cylinder along z-axis
4) Rectangle centered on z-axis

Default is the standard gaussian distribution.  Otherwise user selects from
a menu of options.  The gaussian width parameters are used to define the
parameters of each of the four "flat" distributions.  Their meaning are specified
in the header file.

### CODE REVIEW

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx
Comment (plexoos) [inline StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx]: I think the use of `this->` for `mVx`, `mVy`, and `mVz` is critical here

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx
Comment (plexoos) [inline StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx]: Why do you use lambdas instead of regular functions?
- The functions are not very small, you declare new variables and even have loops inside lambdas
- It is a bit hard to see the signature of the function if it is not declared as `Out myfunc(Inp1, Inp2, ...);`
- You also capture by reference the entire object, which may lead to hard-to-find errors if there is a mistake
- If one day you decide to write a unit test for your random distributions, it is much easier to do for a regular function

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.h
Comment (plexoos) [inline StRoot/StarGenerator/BASE/StarPrimaryMaker.h]: Are you sure you need this setter? Is there a point in switching to a different vertex distribution after the maker is initialized? This option can be added as an argument to the constructor with a default value.

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.h
Comment (plexoos) [inline StRoot/StarGenerator/BASE/StarPrimaryMaker.h]: Please consider using an enum instead of a collection of strings.

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx
Comment (plexoos) [inline StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx]: This map is defined outside of the maker but is initialized in the maker. This is an undefined behavior when two instances are created (as in MT :) ) even though they do the same thing.

Comment (klendathu2k) [reply StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx]: To be fixed in the next commit / push.

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx
Comment (klendathu2k) [reply StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx]: Mainly because it was less boilerplate than binding to a member function.  However, this is changed in the next set of commits / push.

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.h
Comment (klendathu2k) [reply StRoot/StarGenerator/BASE/StarPrimaryMaker.h]: At some point I would like to provide a method which adds distributions to the menu.   Enum would require a recompile of the maker.

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.h
Comment (klendathu2k) [reply StRoot/StarGenerator/BASE/StarPrimaryMaker.h]: The StMaker framework design discourages setting maker parameters as constructor arguments.... but you are correct that a setter is not quite right here.  I should set an string attribute.  To be done...

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.h
Comment (plexoos) [inline StRoot/StarGenerator/BASE/StarPrimaryMaker.h]: Per coding guidelines: vertexFunctionMap -> mVertexFunctionMap

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx
Comment (plexoos) [inline StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx]: For consistency 'this' may be dropped as in the other cases

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx
Comment (plexoos) [inline StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx]: You may want to verify the user input. If non-existent key is provided it'll probably yield (0,0,0) for vertex position (maybe OK) or crash? (not OK)

**Review by plexoos — APPROVED** [review]
LGTM. Only minor comments from my side for you consideration. The class seems to support the slope in XZ and YZ but it is not implemented in the random distributions. An easy fix to add or may be in the next iteration...

**Comment by klendathu2k** [comment]
> 

FYI... the beamline constraint (or user specified slope) is applied after the vertex is sampled (via the RotateBeamline method).

File: StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx
Comment (klendathu2k) [reply StRoot/StarGenerator/BASE/StarPrimaryMaker.cxx]: Done.  Default will be gaussian.  ... with a warning issued.

**Comment by plexoos** [comment]
Jason, don't worry about the failed job. It is probably something I introduced by updating the star-test-data container with new test files. Investigating...

I think there is no point in waiting for other approvals so, let's merge this soon.

**Comment by klendathu2k** [comment]
Agreed it is ready to move (do_events.C shouldn't have any dependence on 
StarGenerator... that
macro is at least 10 years older than StarGen).  Thanks.

Jason

**Pull request merged**
